core/src/style/polylineStyle.cpp

```diff
@@ -98,164 +227,135 @@ double widthMeterToPixel(int _zoom, double _tileSize, double _width) {
+    Builders::buildPolyLine(_line, m_builder);
 
-    return p;
+    _mesh.indices.insert(_mesh.indices.end(),
+                         m_builder.indices.begin(),
+                         m_builder.indices.end());
+
+    _mesh.offsets.emplace_back(m_builder.indices.size(),
+                               m_builder.numVertices);
+
+    m_builder.clear();
 }
 
-void PolylineStyle::buildMesh(const Line& _line, Parameters& _params, VboMesh& _mesh) const {
+void PolylineStyleBuilder::addMesh(const Line& _line, const Parameters& _params) {
 
-    std::vector<PolylineVertex> vertices;
+    m_builder.cap = _params.fill.cap;
+    m_builder.join = _params.fill.join;
+    m_builder.keepTileEdges = _params.keepTileEdges;
 
-    std::vector<uint16_t> fillIndices;
-    std::vector<PolylineVertex> fillVertices;
+    buildLine(_line, _params.fill, m_meshData[0]);
 
-    GLuint color = _params.fill.color;
-    float width = _params.fill.width;
-    float slope = _params.fill.slope;
-    float order = _params.fill.order;
+    if (!_params.outlineOn) { return; }
 
-    if (width <= 0.f && slope <= 0.f ) { return; }
+    if (_params.stroke.cap != _params.fill.cap ||
+        _params.stroke.join != _params.fill.join) {
+        // need to re-triangulate with different cap and/or join
+        m_builder.cap = _params.stroke.cap;
+        m_builder.join = _params.stroke.join;
 
-    PolyLineBuilder builder {
-        [&](const glm::vec3& coord, const glm::vec2& normal, const glm::vec2& texCoord) {
-            glm::vec3 position = glm::vec3{coord.x, coord.y, _params.height};
-            vertices.push_back({position, order, texCoord, normal, { width, slope }, color});
-        },
-        [&](size_t sizeHint){ vertices.reserve(sizeHint); },
-        _params.fill.cap,
-        _params.fill.join,
-        _params.keepTileEdges
-    };
+        buildLine(_line, _params.stroke, m_meshData[1]);
 
-    Builders::buildPolyLine(_line, builder);
-    vertices.swap(fillVertices);
-    builder.indices.swap(fillIndices);
-    // clear buider vertices for a new build call
-    builder.clear();
-
-    if (_params.outlineOn && (_params.outline.width > 0.f || _params.outline.slope > 0.f)) {
-
-        // Must update existing variables, they are captured (and used) by the builder
-        color = _params.outline.color;
-        width += _params.outline.width;
-        slope += _params.outline.slope;
-        order = std::min(_params.outline.order, _params.fill.order) - 0.5f;
-
-        if (_params.outline.cap != _params.fill.cap || _params.outline.join != _params.fill.join) {
-            // need to re-triangulate with different cap and/or join
-            builder.cap = _params.outline.cap;
-            builder.join = _params.outline.join;
-            Builders::buildPolyLine(_line, builder);
-        } else {
-            // re-use indices and positions from original line
+    } else {
+        auto& fill = m_meshData[0];
+        auto& stroke = m_meshData[1];
```

Comment:
From the discussion in #502 I understood that the goal was to draw all outlines first. 

Though thinking about it now, it would make sense to switch the order unless blending mode is overlay to reduce overdraw.